libs/providers/rocketflag/src/lib/rocketflag-provider.ts

```diff
@@ -0,0 +1,141 @@
+import { EventEmitter } from 'events';
+
+export interface UserContext {
+  cohort?: string;
+}
+
+export interface FlagStatus {
+  enabled: boolean;
+}
+
+interface RocketFlagClient {
+  getFlag(flagKey: string, userContext: UserContext): Promise<FlagStatus>;
+}
+
+/**
+ * A helper function to fetch a flag, update the cache, and notify OpenFeature of changes.
+ * It's defined within the factory's scope to access the client, cache, emitter, and logger.
+ */
+const fetchFlagAndUpdateCache = (
+  // Parameters required for the fetch operation
+  flagKey: string,
+  defaultValue: boolean,
+  context: EvaluationContext,
+  cacheKey: string,
+  // State managed by the factory's closure
+  client: RocketFlagClient,
+  cache: Map<string, ResolutionDetails<boolean>>,
+  emitter: EventEmitter,
+  logger?: Logger,
+) => {
+  const userContext: UserContext = {};
+  const { targetingKey } = context;
+
+  if (targetingKey && typeof targetingKey === 'string' && targetingKey !== '') {
+    userContext.cohort = targetingKey;
+  }
+
+  client
+    .getFlag(flagKey, userContext)
+    .then((flagStatus) => {
+      const details: ResolutionDetails<boolean> = {
+        value: flagStatus.enabled,
+        reason: userContext.cohort ? StandardResolutionReasons.TARGETING_MATCH : StandardResolutionReasons.DEFAULT,
+      };
+      cache.set(cacheKey, details);
+      emitter.emit(ProviderEvents.ConfigurationChanged, { flagsChanged: [flagKey] });
+      logger?.debug(`Successfully fetched flag: ${flagKey}`);
+    })
+    .catch((error: unknown) => {
+      const err = error instanceof Error ? error : new Error(String(error));
+      const details: ResolutionDetails<boolean> = {
+        value: defaultValue,
+        reason: StandardResolutionReasons.ERROR,
+        errorCode: ErrorCode.GENERAL,
+        errorMessage: err.message,
+      };
+      cache.set(cacheKey, details);
+      emitter.emit(ProviderEvents.ConfigurationChanged, { flagsChanged: [flagKey] });
+      logger?.error(`Error fetching flag: ${flagKey}`, err);
+    });
+};
+
+/**
+ * Creates a functional OpenFeature provider for RocketFlag.
+ * This provider resolves boolean flags from the RocketFlag service.
+ *
+ * @param {RocketFlagClient} client - An instance of the RocketFlag client.
+ * @returns {Provider & EventEmitter} A provider instance that can be used with the OpenFeature SDK.
+ */
+export function createRocketFlagProvider(client: RocketFlagClient): Provider & EventEmitter {
+  const emitter = new EventEmitter();
+  const cache = new Map<string, ResolutionDetails<boolean>>();
+  let logger: Logger | undefined;
+
+  // Define the provider's logic in a plain object.
+  const providerLogic = {
+    metadata: {
+      name: 'RocketFlagProvider',
+    },
+    runsOn: 'client' as const,
+    hooks: [],
+
+    initialize: async (_context: EvaluationContext, initLogger?: Logger): Promise<void> => {
+      logger = initLogger;
+      logger?.debug('Initialising RocketFlagProvider...');
+    },
```

Comment:
Is it possible in Rocketflag to bulk evaluate or fetch a copy of the ruleset for local evaluation? The current implementation would likely work fine in React where the components would rerender once the flag has finished evaluating but the experience wouldn't be great if the user is using the web SDK directly

Reply:
No, RocketFlag doesn't have a bulk evaluate api that you're referring to.